### PR TITLE
Allow 'ref' and 'in' parameters to be null-checked

### PR DESCRIPTION
**Pull request by RikkiGibson**
Related to #36024
Related to #58699

Please take a look @cston @chsienki @CyrusNajmabadi

### CODE REVIEW

**Review by CyrusNajmabadi — APPROVED** [review]
ide side lgtm

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 1)

File: src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs [hunk 1]
```diff
@@ -551,17 +551,43 @@ public C(string s!!)
...
+    public C(" + refKind + @" string s!!)
+    {
+    }
+}",
+                LanguageVersion = LanguageVersionExtensions.CSharpNext
+            }.RunAsync();
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseIsNullCheck)]
+        public async Task TestOutParameter()
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:583]: It's more common that we assert the expected compiler diagnostics here, ie, remove `CompilerDiagnostics = ...` below.

File: src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs [hunk 2]
```diff
@@ -551,17 +551,43 @@ public C(string s!!)
...
+        {
+            var testCode = @"
+using System;
+
+class C
+{
+    public C(out string s)
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:590]:
```suggestion
    public {|CS0177:C|}(out string s)
```

File: src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs [hunk 3]
```diff
@@ -551,17 +551,43 @@ public C(string s!!)
...
     {
         if (s is null)
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:592]:
```suggestion
        if ({|CS0269:s|} is null)
```

**Review by cston — APPROVED** [review]

File: src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Analyzers/CSharp/Tests/UseParameterNullChecking/UseParameterNullCheckingTests.cs:583]: I think this is a reasonable suggestion, and would be happy to take a change that makes this happen in a follow up.

**Pull request merged**